File: Part 4 - Clustering/Section 25 - Hierarchical Clustering/my_hc.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import scipy.cluster.hierarchy as sch
from sklearn.cluster import AgglomerativeClustering

# %matplotlib qt

# Importing dataset
dataset = pd.read_csv("Mall_Customers.csv")
X = dataset.iloc[:, [3,4]].values

# Creating dendrogram to choose amount of clusters
dendrogram = sch.dendrogram(sch.linkage(X, method = "ward"))
plt.show()

# Choice: 5, perform clustering
hc = AgglomerativeClustering(n_clusters = 5, affinity = "euclidean", 
                             linkage = "ward")
y_pred = hc.fit_predict(X)

plt.scatter(X[y_pred == 0, 0], X[y_pred == 0, 1], c = "red", s = 100)
plt.scatter(X[y_pred == 1, 0], X[y_pred == 1, 1], c = "green", s = 100)
plt.scatter(X[y_pred == 2, 0], X[y_pred == 2, 1], c = "blue", s = 100)
plt.scatter(X[y_pred == 3, 0], X[y_pred == 3, 1], c = "grey", s = 100)
plt.scatter(X[y_pred == 4, 0], X[y_pred == 4, 1], c = "purple", s = 100)
plt.xlabel("Income")
plt.ylabel("Spending")
plt.show()

File: Part 4 - Clustering/Section 25 - Hierarchical Clustering/my_hc.R
# Import the dataset
dataset <- read.csv("Mall_Customers.csv")
X <- dataset[4:5]

# Using dendrogram to choose amount of clusters
dendrogram = hclust(dist(X, method = "euclidean"), method = "ward.D")
plot(dendrogram)

# The optimal is 5
# Create 5 clusters
clustered <- hclust(dist(X, method = "euclidean"), method = "ward.D")
clusters <- cutree(clustered, k = 5)

clusplot(X,
         clusters,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 1,
         plotchar = FALSE,
         span = TRUE,
         main = "Cluster plot",
         xlab = 'Income',
         ylab = 'Spending')